### ISSUE TITLE
new mnemonic/seed version wishlist

### ISSUE THREAD

**Issue by SomberNight** [spesmilo/electrum#5782]
ref: https://electrum.readthedocs.io/en/latest/seedphrase.html

For future seed versions, it would be nice if we could
1. encode a birthday
    - At least Electrum Personal Server would benefit if we knew where to start scanning the blockchain. (Although we would also need a mechanism to expose this to EPS somehow; maybe encode it together with xpub?) @chris-belcher 
    - I don't see how this could be easily done without changing the scheme substantially, as currently all the info we encode in the seed besides the entropy (only the version number) is encoded via bruteforce, making it difficult to add any new fields.
2. use stronger KDF
    - currently we use the same KDF as BIP39, which is really weak by any standard

**Comment by chris-belcher** [comment]
Yes encoding the birthday into the MPK is probably the best method.

**Comment by chris-belcher** [comment]
It's possible that birthday for seed might not be useful long term, because from what I see full node software is moving towards pruning and scanning the UTXO set. In the long term Bitcoin Core might come with AssumeUTXO which avoids downloading the most of the blockchain entirely but only has the UTXO set. Software like btcpayserver and specter wallet either already do or are in the process of implementing a feature where they come with someone's pre-sync'd and pruned Bitcoin Core data directory.

If we really are moving in that direction, then recovering a wallet from seed phrase with the user's own full node will only ever give the user their unspent coins, not their history and spent transactions.


An entirely unrelated idea of new features to the seed phrase format: multisig parameters. It would be nice that when a user types in a seed phrase then Electrum can tell them "this seed phrase is part of a 3-of-5 multisig wallet, to spend the coins you must have 2 other seed phrases".

**Comment by SomberNight** [comment]
>> encode a birthday 
>
> I don't see how this could be easily done without changing the scheme substantially, as currently all the info we encode in the seed besides the entropy (only the version number) is encoded via bruteforce, making it difficult to add any new fields.

Just to clarify, my point here is that with the current design, it is difficult to store a lot of information in the seed because we are grinding seeds until we find one which has a hash that encodes what we want.

The current scheme (electrum2) (for 2fa and for segwit seeds) grinds 12 bits using hmac-sha512 (so ~4k expected attempts). These 12 bits have a two-fold function:
- encode that e.g. p2wpkh scripts are used with an implicit derivation path
- it is a checksum for the seed as almost all values are invalid

Atm there are only four defined script types, and the rest of the 8-12 bits is practically a checksum:
https://github.com/spesmilo/electrum/blob/9c4807644bf8bb0d2b777e9d02573e9b94db7c0f/electrum/version.py#L6-L10

Note that grinding 12 bits already takes in the ballpark of 1 second on a slow Android phone, and I don't think it can be sped up a lot by changing the hash function.

I believe the sole reason we are grinding is to be wordlist-independent. That is, no wordlist is needed to decode a seed.

-----

I was thinking how we could encode more than a few bits of information, while keeping the no-wordlist-needed feature.
(would be trivial to encode as much as we wanted if we fixed a wordlist)

One idea I have is that we could consider the seed a concatenation of sub-seeds; i.e. taken to the extreme, we could grind each word. This would imply that we started assuming some kind of structure to electrum seeds though, in this case that **there are words** (i.e. whitespace-separated string). That is, no more `9dk` seed (which is a valid "segwit"-type electrum2 seed, that has some testnet history of mine :P).

For example, we could grind 8 bits on each word. With a 12-word seed, that allows storing 96 bits of arbitrary data (minus checksum). Just like with the electrum2 seed format, the entropy could be extracted by using a different hash function on the full string/seed.

Caveats:
- Obviously, the more custom data we store, the less the useful entropy of the seed is. That is, with a 12-word seed generated from a 2048-long wordlist, we get 132 bits of space, which is the ~sum of the stored custom data and the entropy, i.e. if we stored 96 bits of custom data (birthdate, multisig parameters, etc), only 36 bits of entropy is left. So, unsurprisingly, the more custom data we want to store, the longer the seed needs to be.
- Decoding needs no wordlist, but seed-generation obviously needs one for practical reasons (same as now). Grinding on the word level imposes new restrictions on this wordlist. E.g. with a 2048-long list, it seems reasonable that all 8-bit values occur at least once but this would need to be checked and kept in mind at design-time.
    - However, what we actually would want is a uniform distribution of the grinded values across the wordlist, e.g. 3-bit worth of words for every 8-bit grinded value, i.e. in the 2048-long list, 8 words for every 8-bit grinded value

If there arises a legitimate need to store "lots of" custom data in the seed, it seems significantly easier to drop the "no-wordlist-needed" feature.

**Comment by ecdsa** [comment]
> Just to clarify, my point here is that with the current design, it is difficult to store a lot of information in the seed because we are grinding seeds until we find one which has a hash that encodes what we want.

I do not understand your concern. Electrum's seed version system does not assume anything about how the seed is constructed. 
If we introduce a fixed wordlist, we may design something like:

```
 binary_seed = entropy || extra_data
 seed_phrase = bidirectional_encoding(binary_seed)
 seed_version = first_bits_of(hmac_sha_512("Seed version", seed_phrase))
```
We would grind the entropy part of the binary_seed, and that would be compatible with the existing system. The only drawback is that we need to introduce a fixed word list for the encoding. However, that seems to be what we have been doing anyway in order to facilitate typing on Android.

**Comment by SomberNight** [comment]
@ecdsa

> I do not understand your concern

I was trying to come up with a design without fixing a wordlist.

like I said:

>  (would be trivial to encode as much as we wanted if we fixed a wordlist)

**Comment by SomberNight** [comment]
note: re derivation path, we should make it depend on `BIP44_COIN_TYPE`

https://github.com/spesmilo/electrum/blob/be084dd44b76d790592e49413b50e9b34d757086/electrum/constants.py#L102

This would make keys differ between mainnet and testnet, in case some people reuse their seed.
More realistically than testnet though, some people might be reusing their electrum seed with forks of electrum (e.g. for LTC).

**Comment by SomberNight** [comment]
Edited the OP to include two recent ideas from discussion with @ecdsa:

- encode gap limit into seed, or maybe make it only changeable using OP_RETURNs (or similar info recoverable from chain data)
- support for additional script templates or derivation paths (perhaps miniscript) could be added later on if we specified a generic way for it via OP_RETURNs
    - e.g. scan wallet history for an OP_RETURN that decrypts with an HD-derived key. The decrypted payload could be something complex such as a miniscript template, which we should also start to scan for, or e.g. just a magic version number. If the version number is known to the given version of the software, start deriving addresses according to what is specified, or if it is an unknown (future or past) version number, the software can display a warning to the user.
    - basically this would give us the generality and extensibility of bip39/44, but in a way where the software knows about unknown derivation paths/scripts
